File: Data Structure & Algorithm/Data Structure & Algorithm [Old]/Models/Tree/Binary Tree/Tree Restore/Construct Binary Tree from Preorder and Inorder Traversal/Construct Binary Tree from Inorder and Postorder Traversal.py
from typing import List, Optional


class TreeNode:

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    def buildTree(self, preorder: List[int],
                  inorder: List[int]) -> Optional[TreeNode]:

        def recursive(preorder: List[int],
                      inorder: List[int]) -> Optional[TreeNode]:
            if len(preorder) <= 0 or len(inorder) <= 0:
                return None

            currentVal = preorder.pop(0)
            currentIndex = inorder.index(currentVal)
            node = TreeNode(currentVal)

            node.left = recursive(preorder, inorder[0:currentIndex])
            node.right = recursive(preorder, inorder[currentIndex + 1:])

            return node

        return recursive(preorder, inorder)


# test
def levelOrder(root: TreeNode | None) -> List[int]:
    res = []
    queue = []

    if root == None:
        return res

    queue.append(root)
    while (len(queue) > 0):
        size = len(queue)
        for _ in range(size):
            current = queue.pop(0)
            res.append(current.val)
            if (current.left):
                queue.append(current.left)
            if (current.right):
                queue.append(current.right)
    return res


preorder = [3, 9, 20, 15, 7]
inorder = [9, 3, 15, 20, 7]
solution = Solution()
res = solution.buildTree(preorder, inorder)
printTree = levelOrder(res)
print(printTree)

File: Data Structure & Algorithm/Models/Tree/Binary Tree/Tree Restore/Construct Binary Tree from Preorder and Inorder Traversal/Construct Binary Tree from Inorder and Postorder Traversal.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func recursive(preorder []int, inorder []int) *TreeNode {
	if len(preorder) <= 0 || len(inorder) <= 0 {
		return nil
	}

	currentVal := preorder[0]
	currentIndex := 0
	for i, val := range inorder {
		if val == currentVal {
			currentIndex = i
			break
		}
	}

	node := &TreeNode{currentVal, nil, nil}
	node.Left = recursive(preorder[1:currentIndex+1], inorder[0:currentIndex])
	node.Right = recursive(preorder[currentIndex+1:], inorder[currentIndex+1:])
	return node
}

func buildTree(preorder []int, inorder []int) *TreeNode {
	return recursive(preorder, inorder)
}

// test
func levelOrder(root *TreeNode) []int {
	res := []int{}
	queue := []*TreeNode{}

	if root == nil {
		return res
	}

	queue = append(queue, root)
	for len(queue) > 0 {
		size := len(queue)
		for i := 0; i < size; i++ {
			current := queue[0]
			queue = queue[1:]
			res = append(res, current.Val)
			if current.Left != nil {
				queue = append(queue, current.Left)
			}
			if current.Right != nil {
				queue = append(queue, current.Right)
			}
		}
	}

	return res
}

func main() {
	preorder := []int{3, 9, 20, 15, 7}
	inorder := []int{9, 3, 15, 20, 7}
	res := buildTree(preorder, inorder)
	printTree := levelOrder(res)
	fmt.Println(printTree)
}
